src/devices/Mcp25xxx/Mcp25xxx.cs

```diff
@@ -0,0 +1,407 @@
+        /// <param name="gpioController">
+        /// The GPIO controller for defined external pins. If not specified, the default controller will be used.
+        /// </param>
+        public Mcp25xxx(
+            SpiDevice spiDevice,
+            int reset = -1,
+            int tx0rts = -1,
+            int tx1rts = -1,
+            int tx2rts = -1,
+            int interrupt = -1,
+            int rx0bf = -1,
+            int rx1bf = -1,
+            int clkout = -1,
+            IGpioController gpioController = null)
+        {
+            _spiDevice = spiDevice;
+
+            _reset = reset;
+            _tx0rts = tx0rts;
+            _tx1rts = tx1rts;
+            _tx2rts = tx2rts;
+            _interrupt = interrupt;
+            _rx0bf = rx0bf;
+            _rx1bf = rx1bf;
+            _clkout = clkout;
+
+            // Only need master controller if there are external pins provided.
+            if (_reset != -1 ||
+                _tx0rts != -1 ||
+                _tx1rts != -1 ||
+                _tx2rts != -1 ||
+                _interrupt != -1 ||
+                _rx0bf != -1 ||
+                _rx1bf != -1 ||
+                _clkout != -1)
+            {
+                _gpioController = gpioController ?? new GpioController();
+
+                if (_reset != -1)
+                {
+                    _gpioController.OpenPin(_reset, PinMode.Output);
+                    ResetPin = PinValue.Low;
+                }
+
+                if (_tx0rts != -1)
+                {
+                    _gpioController.OpenPin(_tx0rts, PinMode.Output);
+                }
+
+                if (_tx1rts != -1)
+                {
+                    _gpioController.OpenPin(_tx1rts, PinMode.Output);
+                }
+
+                if (_tx2rts != -1)
+                {
+                    _gpioController.OpenPin(_tx2rts, PinMode.Output);
+                }
+
+                if (_interrupt != -1)
+                {
+                    _gpioController.OpenPin(_interrupt, PinMode.Input);
+                }
+
+                if (_rx0bf != -1)
+                {
+                    _gpioController.OpenPin(_rx0bf, PinMode.Input);
+                }
+
+                if (_rx1bf != -1)
+                {
+                    _gpioController.OpenPin(_rx1bf, PinMode.Input);
+                }
+
+                if (_clkout != -1)
+                {
+                    _gpioController.OpenPin(_clkout, PinMode.Input);
+                }
+            }
+        }
+
+        internal void VerifyPinConfigured(int pin, string name)
+        {
+            if (pin == -1)
+            {
+                throw new InvalidOperationException($"No {name} pin configured.");
+            }
+        }
+
+        /// <summary>
+        /// Writes a value to Tx0RTS pin.
+        /// </summary>
+        public PinValue Tx0RtsPin
+        {
+            set
+            {
+                VerifyPinConfigured(_tx0rts, nameof(Tx0RtsPin));
```

Comment:
optional: since you are repeating the logic here, might be worh to create something like: `VerifyingWrite(pin, value)` which will both validate and write. Not sure if argument name is that much useful considering it will be in the top of the stack trace but you can add it as well

Reply:
actually _gpioController.Write should already verify so perhaps might be worth to just remove verification at all... (no strong opinion either way)

Reply:
I see your point on the verify within _gpioController.Write, but this is more of the actual pin.  I like having the check using name as it will show/help to know the name to review/troubleshoot.  Although, this issue will probably not occur much so I'll change to only use _gpioController.Write.  This will slightly help performance on each pin write, as well.   Will be in next commit.

Reply:
fair enough